packages/dataviews/src/test/dataform.tsx

```diff
@@ -0,0 +1,346 @@
+	title: {
+		view: () =>
+			screen.getByRole( 'button', {
+				name: /edit title/i,
+			} ),
+		edit: () =>
+			screen.getByRole( 'textbox', {
+				name: /title/i,
+			} ),
+	},
+	author: {
+		view: () =>
+			screen.getByRole( 'button', {
+				name: /edit author/i,
+			} ),
+		edit: () =>
+			screen.queryByRole( 'combobox', {
+				name: /author/i,
+			} ),
+	},
+	order: {
+		view: () =>
+			screen.getByRole( 'button', {
+				name: /edit order/i,
+			} ),
+		edit: () =>
+			screen.getByRole( 'spinbutton', {
+				name: /order/i,
+			} ),
+	},
+};
+
+describe( 'DataForm component', () => {
+	describe( 'in regular mode', () => {
+		it( 'should display fields', () => {
+			render(
+				<Dataform
+					onChange={ () => void 0 }
+					fields={ fields }
+					form={ form }
+					data={ data }
+				/>
+			);
+
+			expect( fieldsSelector.title.edit() ).toBeInTheDocument();
+			expect( fieldsSelector.order.edit() ).toBeInTheDocument();
+			expect( fieldsSelector.author.edit() ).toBeInTheDocument();
+		} );
+
+		it( 'should render custom Edit component', () => {
+			const fieldsWithCustomEditComponent = fields.map( ( field ) => {
+				if ( field.id === 'title' ) {
+					return {
+						...field,
+						Edit: () => {
+							return <span>This is the Title Field</span>;
+						},
+					};
+				}
+				return field;
+			} );
+
+			render(
+				<Dataform
+					onChange={ () => void 0 }
+					fields={ fieldsWithCustomEditComponent }
+					form={ form }
+					data={ data }
+				/>
+			);
+
+			const titleField = screen.getByText( 'This is the Title Field' );
+			expect( titleField ).toBeInTheDocument();
+		} );
+
+		it( 'should call onChange with the correct value for each typed character', async () => {
+			const onChange = jest.fn();
+			render(
+				<Dataform
+					onChange={ onChange }
+					fields={ fields }
+					form={ form }
+					data={ { ...data, title: '' } }
+				/>
+			);
+
+			const titleInput = fieldsSelector.title.edit();
+			const user = userEvent.setup();
+			await user.clear( titleInput );
+			expect( titleInput ).toHaveValue( '' );
+			const newValue = 'Hello folks!';
+			await user.type( titleInput, newValue );
+			expect( onChange ).toHaveBeenCalledTimes( newValue.length );
+			for ( let i = 0; i < newValue.length; i++ ) {
+				expect( onChange ).toHaveBeenNthCalledWith( i + 1, {
+					title: newValue[ i ],
+				} );
+			}
+		} );
+
+		it( 'should wrap fields in HStack when labelPosition is set to side', async () => {
+			const { container } = render(
+				<Dataform
+					onChange={ () => void 0 }
+					fields={ fields }
+					form={ { ...form, labelPosition: 'side' } }
+					data={ data }
+				/>
+			);
+
+			expect(
+				// It is used here to ensure that the fields are wrapped in HStack. This happens when the labelPosition is set to side.
+				// eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
+				container.querySelectorAll( "[data-wp-component='HStack']" )
+			).toHaveLength( 3 );
+		} );
```

Comment:
This kind test is too implementation-specific in my view. We're testing a specific HTML output that we may need to change. Personally, I find tests more useful when they test the API. Don't feel you need to remove it, but I find it a bit fragile.

Reply:
💯 agree with you, but I didn't find a better way to “test” this behavior.